File: Java/Sample_Source/ContentModification/ExtendedGraphicStates/src/com/datalogics/PDFL/Samples/ExtendedGraphicStates.java
package com.datalogics.pdfl.samples.ContentModification.ExtendedGraphicStates;

/*
 * 
 * The Graphics State is an internal data structure in a PDF file that holds the parameters that describe graphics
 * within that file. These parameters define how individual graphics are presented on the page. Adobe Systems introduced
 * the Extended Graphic State to expand the original Graphics State data structure, providing space to define and store
 * more data objects within a PDF.
 * 
 * This sample program shows how to use the Extended Graphic State object to add graphics parameters to an image.
 * 
 * For more detail see the description of the ExtendedGraphicState sample program on our Developer's site, 
 * http://dev.datalogics.com/adobe-pdf-library/sample-program-descriptions/java-sample-programs/adding-text-and-graphic-elements#extendedgraphicstates
 * 
 * Copyright (c) 2007-2017, Datalogics, Inc. All rights reserved.
 *
 * For complete copyright information, refer to:
 * http://dev.datalogics.com/adobe-pdf-library/license-for-downloaded-pdf-samples/
 *
 */
import com.datalogics.PDFL.BlendMode;
import com.datalogics.PDFL.Color;
import com.datalogics.PDFL.Content;
import com.datalogics.PDFL.Document;
import com.datalogics.PDFL.ExtendedGraphicState;
import com.datalogics.PDFL.Font;
import com.datalogics.PDFL.FontCreateFlags;
import com.datalogics.PDFL.GraphicState;
import com.datalogics.PDFL.Image;
import com.datalogics.PDFL.Library;
import com.datalogics.PDFL.Matrix;
import com.datalogics.PDFL.Page;
import com.datalogics.PDFL.Rect;
import com.datalogics.PDFL.SaveFlags;
import com.datalogics.PDFL.Text;
import com.datalogics.PDFL.TextRun;
import com.datalogics.PDFL.TextState;
import java.util.EnumSet;

public class ExtendedGraphicStates {

    static void blendPage(Document doc, Image foregroundImage, Image backgroundImage) {

        // This section demonstrates all the Blend Modes one can achieve
        // by setting the BlendMode property to each of the 16 enumerations
        // on a foreground image over a background image, and
        // plopping all these images on a single page.
        double height = 792;
        double width = 612;

        Rect pageRect = new Rect(0, 0, width, height);
        Page docpage = doc.createPage(doc.getNumPages() - 1, pageRect);
        Text t = new Text();
        GraphicState gsText = new GraphicState();
        gsText.setFillColor(new Color(0, 0, 1.0));
        TextState ts = new TextState();
        Font f = null;
        TextRun tr = null;
        Image individualForegroundImage = null;
        Image individualBackgroundImage = null;
        Matrix m = null;
        GraphicState gs = null;
        ExtendedGraphicState xgs = null;
        Content pgContent = null;

        try {
            pgContent = docpage.getContent();
            f = new Font("CourierStd", EnumSet.of(FontCreateFlags.EMBEDDED, FontCreateFlags.SUBSET));

            for (int i = 0; i < 16; i++) {
                try {
                    individualForegroundImage = foregroundImage.clone();
                    individualBackgroundImage = backgroundImage.clone();

                    gs = individualForegroundImage.getGraphicState();
                    individualForegroundImage.scale(0.125, 0.125);
                    individualForegroundImage.translate(800, 200 + height * (7 - i));
                    individualBackgroundImage.scale(0.125, 0.125);
                    individualBackgroundImage.translate(800, 200 + height * (7 - i));

                    // Halfway through, create 2nd column by shifting over and
                    // up
                    if (i > 7) {
                        individualForegroundImage.translate(2400, height * 8);
                        individualBackgroundImage.translate(2400, height * 8);
                    }

                    m = new Matrix();
                    if (i > 7) {
                        m = m.translate(480, 750 - ((i - 8) * 100)); // second
                                                                     // column
                    } else {
                        m = m.translate(180, 750 - (i * 100)); // first column
                    }
                    m = m.scale(12.0, 12.0);

                    xgs = new ExtendedGraphicState();
                    if (i == 0) {
                        xgs.setBlendMode(BlendMode.NORMAL);
                        tr = new TextRun("Normal", f, gsText, ts, m);
                    } else if (i == 1) {
                        xgs.setBlendMode(BlendMode.MULTIPLY);
                        tr = new TextRun("Multiply", f, gsText, ts, m);
                    } else if (i == 2) {
                        xgs.setBlendMode(BlendMode.SCREEN);
                        tr = new TextRun("Screen", f, gsText, ts, m);
                    } else if (i == 3) {
                        xgs.setBlendMode(BlendMode.OVERLAY);
                        tr = new TextRun("Overlay", f, gsText, ts, m);
                    } else if (i == 4) {
                        xgs.setBlendMode(BlendMode.DARKEN);
                        tr = new TextRun("Darken", f, gsText, ts, m);
                    } else if (i == 5) {
                        xgs.setBlendMode(BlendMode.LIGHTEN);
                        tr = new TextRun("Lighten", f, gsText, ts, m);
                    } else if (i == 6) {
                        xgs.setBlendMode(BlendMode.COLOR_DODGE);
                        tr = new TextRun("Color Dodge", f, gsText, ts, m);
                    } else if (i == 7) {
                        xgs.setBlendMode(BlendMode.COLOR_BURN);
                        tr = new TextRun("Color Burn", f, gsText, ts, m);
                    } else if (i == 8) {
                        xgs.setBlendMode(BlendMode.HARD_LIGHT);
                        tr = new TextRun("Hard Light", f, gsText, ts, m);
                    } else if (i == 9) {
                        xgs.setBlendMode(BlendMode.SOFT_LIGHT);
                        tr = new TextRun("SoftLight", f, gsText, ts, m);
                    } else if (i == 10) {
                        xgs.setBlendMode(BlendMode.DIFFERENCE);
                        tr = new TextRun("Difference", f, gsText, ts, m);
                    } else if (i == 11) {
                        xgs.setBlendMode(BlendMode.EXCLUSION);
                        tr = new TextRun("Exclusion", f, gsText, ts, m);
                    } else if (i == 12) {
                        xgs.setBlendMode(BlendMode.HUE);
                        tr = new TextRun("Hue", f, gsText, ts, m);
                    } else if (i == 13) {
                        xgs.setBlendMode(BlendMode.SATURATION);
                        tr = new TextRun("Saturation", f, gsText, ts, m);
                    } else if (i == 14) {
                        xgs.setBlendMode(BlendMode.COLOR);
                        tr = new TextRun("Color", f, gsText, ts, m);
                    } else if (i == 15) {
                        xgs.setBlendMode(BlendMode.LUMINOSITY);
                        tr = new TextRun("Luminosity", f, gsText, ts, m);
                    }
                    t.addRun(tr);

                    pgContent.addElement(individualBackgroundImage);
                    System.out.println("Added background image " + (i + 1) + " to the content.");

                    System.out.println("Set blend mode in extended graphic state.");
                    gs.setExtendedGraphicState(xgs);
                    individualForegroundImage.setGraphicState(gs);

                    pgContent.addElement(individualForegroundImage);
                    System.out.println("Added foreground image " + (i + 1) + " to the content.");

                    pgContent.addElement(t);

                } finally {
                    tr.delete();
                    individualBackgroundImage.delete();
                    individualForegroundImage.delete();
                    gs.delete();
                    xgs.delete();
                    m.delete();
                }
            }

            docpage.updateContent();
            System.out.println("Updated the content on page " + doc.getNumPages() + ".");
        } finally {
            t.delete();
            f.delete();
            gsText.delete();
            ts.delete();
            pgContent.delete();
            docpage.delete();
        }
    }

    /**
     * @param args
     */
    public static void main(String[] args) throws Throwable {
        System.out.println("ExtendedGraphicStates sample:");

        Library lib = new Library();
        Document doc = null;
        Image ducky = null;
        Image rainbow = null;

        try {
            String sInput1 = Library.getResourceDirectory() + "Sample_Input/ducky_alpha.tif";
            String sInput2 = Library.getResourceDirectory() + "Sample_Input/rainbow.tif";
            String sOutput = "ExtendedGraphicStates-out.pdf";
            System.out.println("Initialized the library.");

            doc = new Document();

            if (args.length > 0)
                sInput1 = args[0];
            if (args.length > 1)
                sInput2 = args[1];
            if (args.length > 2)
                sOutput = args[2];
            System.out.println("Input images: " + sInput1 + " and " + sInput2 + 
                  "; will write to " + sOutput);
            ducky = new Image(sInput1, doc);
            rainbow = new Image(sInput2, doc);

            // Ducky on top of rainbow
            blendPage(doc, ducky, rainbow);

            // Rainbow on top of ducky
            blendPage(doc, rainbow, ducky);

            doc.embedFonts();
            doc.save(EnumSet.of(SaveFlags.FULL), sOutput);
            System.out.println("Saved the document " + sOutput);

            // Kill the doc object
            System.out.println("Killed document object.");
        } finally {
            ducky.delete();
            rainbow.delete();
            doc.delete();
            lib.delete();
        }
    }
}

File: DotNET/Sample_Source/ContentModification/ExtendedGraphicStates/ExtendedGraphicStates.cs
using System;
using Datalogics.PDFL;

/*
 * 
 * The Graphics State is an internal data structure in a PDF file that holds the parameters that describe graphics
 * within that file. These parameters define how individual graphics are presented on the page. Adobe Systems introduced
 * the Extended Graphic State to expand the original Graphics State data structure, providing space to define and store
 * more data objects within a PDF.
 * 
 * This sample program shows how to use the Extended Graphic State object to add graphics parameters to an image.
 * 
 * For more detail see the description of the ExtendedGraphicState sample program on our Developer’s site, 
 * http://dev.datalogics.com/adobe-pdf-library/sample-program-descriptions/net-core-sample-programs/adding-text-and-graphic-elements#extendedgraphicstates
 * 
 * Copyright (c) 2007-2020, Datalogics, Inc. All rights reserved.
 *
 * For complete copyright information, refer to:
 * http://dev.datalogics.com/adobe-pdf-library/license-for-downloaded-pdf-samples/
 *
 */

namespace ExtendedGraphicStates
{
    class ExtendedGraphicStates
    {
        static void blendPage(Document doc, Image foregroundImage, Image backgroundImage)
        {
            double height = 792;
            double width = 612;
            Rect pageRect = new Rect(0, 0, width, height);
            Page docpage = doc.CreatePage(doc.NumPages - 1, pageRect);

            // This section demonstrates all the Blend Modes one can achieve
            // by setting the BlendMode property to each of the 16 enumerations
            // on a foreground "ducky" over a background rainbow pattern, and 
            // plopping all these images on a single page.

            Text t = new Text();
            Font f;
            try
            {
                f = new Font("Arial", FontCreateFlags.Embedded | FontCreateFlags.Subset);
            }
            catch (ApplicationException ex)
            {
                if (ex.Message.Equals("The specified font could not be found.") &&
                    System.Runtime.InteropServices.RuntimeInformation.IsOSPlatform(System.Runtime.InteropServices
                        .OSPlatform.Linux) &&
                    !System.IO.Directory.Exists("/usr/share/fonts/msttcore/"))
                {
                    Console.WriteLine("Please install Microsoft Core Fonts on Linux first.");
                    return;
                }

                throw;
            }

            GraphicState gsText = new GraphicState();
            gsText.FillColor = new Color(0, 0, 1.0);
            TextState ts = new TextState();

            for (int i = 0; i < 16; i++)
            {
                Image individualForegroundImage = foregroundImage.Clone();
                Image individualBackgroundImage = backgroundImage.Clone();

                GraphicState gs = individualForegroundImage.GraphicState;
                individualForegroundImage.Scale(0.125, 0.125);
                individualForegroundImage.Translate(800, 200 + height * (7 - i));
                individualBackgroundImage.Scale(0.125, 0.125);
                individualBackgroundImage.Translate(800, 200 + height * (7 - i));

                // Halfway through, create 2nd column by shifting over and up
                if (i > 7)
                {
                    individualForegroundImage.Translate(2400, height * 8);
                    individualBackgroundImage.Translate(2400, height * 8);
                }

                docpage.Content.AddElement(individualBackgroundImage);
                Console.WriteLine("Added background image " + (i + 1) + " to the content.");
                docpage.Content.AddElement(individualForegroundImage);
                Console.WriteLine("Added foreground image " + (i + 1) + " to the content.");

                Matrix m = new Matrix();
                if (i > 7)
                    m = m.Translate(480, 750 - ((i - 8) * 100)); // second column
                else
                    m = m.Translate(180, 750 - (i * 100)); // first column
                m = m.Scale(12.0, 12.0);

                ExtendedGraphicState xgs = new ExtendedGraphicState();
                TextRun tr = null;
                if (i == 0)
                {
                    xgs.BlendMode = BlendMode.Normal;
                    tr = new TextRun("Normal", f, gsText, ts, m);
                }
                else if (i == 1)
                {
                    xgs.BlendMode = BlendMode.Multiply;
                    tr = new TextRun("Multiply", f, gsText, ts, m);
                }
                else if (i == 2)
                {
                    xgs.BlendMode = BlendMode.Screen;
                    tr = new TextRun("Screen", f, gsText, ts, m);
                }
                else if (i == 3)
                {
                    xgs.BlendMode = BlendMode.Overlay;
                    tr = new TextRun("Overlay", f, gsText, ts, m);
                }
                else if (i == 4)
                {
                    xgs.BlendMode = BlendMode.Darken;
                    tr = new TextRun("Darken", f, gsText, ts, m);
                }
                else if (i == 5)
                {
                    xgs.BlendMode = BlendMode.Lighten;
                    tr = new TextRun("Lighten", f, gsText, ts, m);
                }
                else if (i == 6)
                {
                    xgs.BlendMode = BlendMode.ColorDodge;
                    tr = new TextRun("Color Dodge", f, gsText, ts, m);
                }
                else if (i == 7)
                {
                    xgs.BlendMode = BlendMode.ColorBurn;
                    tr = new TextRun("Color Burn", f, gsText, ts, m);
                }
                else if (i == 8)
                {
                    xgs.BlendMode = BlendMode.HardLight;
                    tr = new TextRun("Hard Light", f, gsText, ts, m);
                }
                else if (i == 9)
                {
                    xgs.BlendMode = BlendMode.SoftLight;
                    tr = new TextRun("SoftLight", f, gsText, ts, m);
                }
                else if (i == 10)
                {
                    xgs.BlendMode = BlendMode.Difference;
                    tr = new TextRun("Difference", f, gsText, ts, m);
                }
                else if (i == 11)
                {
                    xgs.BlendMode = BlendMode.Exclusion;
                    tr = new TextRun("Exclusion", f, gsText, ts, m);
                }
                else if (i == 12)
                {
                    xgs.BlendMode = BlendMode.Hue;
                    tr = new TextRun("Hue", f, gsText, ts, m);
                }
                else if (i == 13)
                {
                    xgs.BlendMode = BlendMode.Saturation;
                    tr = new TextRun("Saturation", f, gsText, ts, m);
                }
                else if (i == 14)
                {
                    xgs.BlendMode = BlendMode.Color;
                    tr = new TextRun("Color", f, gsText, ts, m);
                }
                else if (i == 15)
                {
                    xgs.BlendMode = BlendMode.Luminosity;
                    tr = new TextRun("Luminosity", f, gsText, ts, m);
                }

                t.AddRun(tr);
                docpage.Content.AddElement(t);
                docpage.UpdateContent();
                Console.WriteLine("Updated the content on page 1.");

                gs.ExtendedGraphicState = xgs;
                individualForegroundImage.GraphicState = gs;
                Console.WriteLine("Set blend mode in extended graphic state.");
            }
        }

        static void Main(string[] args)
        {
            Console.WriteLine("ExtendedGraphicStates Sample:");


            // ReSharper disable once UnusedVariable
            using (Library lib = new Library())
            {
                Console.WriteLine("Initialized the library.");

                String sInput1 = Library.ResourceDirectory + "Sample_Input/ducky_alpha.tif";
                String sInput2 = Library.ResourceDirectory + "Sample_Input/rainbow.tif";
                String sOutput = "ExtendedGraphicStates-out.pdf";

                if (args.Length > 0)
                    sInput1 = args[0];

                if (args.Length > 1)
                    sInput2 = args[1];

                if (args.Length > 2)
                    sOutput = args[2];

                Console.WriteLine("Input files: " + sInput1 + " and " + sInput2 + ". Saving to output file: " +
                                  sOutput);

                Document doc = new Document();

                Image ImageOne = new Image(sInput1, doc);
                Image imageTwo = new Image(sInput2, doc);

                blendPage(doc, ImageOne, imageTwo);

                blendPage(doc, imageTwo, ImageOne);

                doc.EmbedFonts();
                doc.Save(SaveFlags.Full, sOutput);
            }
        }
    }
}
